Add Parking Lot API with add, remove, and list operations

arturoaviles opened the pull request:
The following actions were added to the API:

- Add a car
- Remove a car
- List the parked cars

Other things added:

- Unit tests
- Integration Tests

- Dockerfile and Docker Compose files

arturoaviles committed:
Add Parking Lot API with add, remove, and list operations

aspirin325 commented on app/models/models.py:
```diff
@@ -0,0 +1,151 @@
+		Returns:
+			List[Dict[str, Any]]: List of cars in the Parking Lot.
+		"""
+		return [vars(ticket) for ticket in self.occupied_spots.values()]
```
Let's say we have > 100,000 parked cars, what could be implemented to don't overload the client when requesting the list of parked cars?

arturoaviles replied:
I would implement pagination to retrieve just a limited amount of cars per request. 

I implemented a simple paginator in parking_lot.list_cars. It has: 

- "start" (int): from which parking spot number do you want to retrieve.
- "limit" (int): how many items to retrieve (default=10, max=100)

Another thing that could help is implementing a "filter" flag to return only the defined parts of the ticket.

---
Other possible improvements:

- If a disk database is later implemented to save the data, I would cache the retrieved cars in an in-memory database (redis/memcache) until a write action is made (adding/removing car)

aspirin325 commented on app/models/models.py:
```diff
@@ -0,0 +1,151 @@
+			raise ValueError("Invalid date format") from e
+		return finish - start
+
+	def calculate_fee(self, end_datetime: datetime = datetime.now()) -> None:
+		"""Calculate the fee for the ticket
+
+		Returns:
+			None
+		"""
+		self.finish = end_datetime.strftime(DATE_TIME_FORMAT)
+		delta = self._get_total_time()
+		total_minutes = delta / timedelta(minutes=1)
+
+		if total_minutes <= FREE_MINUTES:
+			self.fee = "0.00"
+			return
+
+		time_frame = TIME_FRAME_HOURS.get(self.tariff)
+		tariff = TIME_FRAME_TARIFFS.get(self.tariff)
+
+		total_time = delta / timedelta(hours=time_frame)
+
+		if not total_time.is_integer():
+			total_time += 1
+
+		self.fee = str(Decimal(tariff) * Decimal(floor(total_time)))
```
What changes are necessary to support new tariffs dynamically ?

arturoaviles replied:
The Parking Lot owner would need an API endpoint to change the tariffs as wanted.

I implemented a /config endpoint to change the parking lot tariffs using a POST request. If the owner wants to know the current tariffs without adding a car, they can be retrieved with the same endpoint by using a GET request instead.

kingdomik commented on app/main.py:
```diff
@@ -0,0 +1,109 @@
+				"status": "error",
+				"description": str(e)
+			},
+			status_code=200
```
Why status 200 if lot is full?

kingdomik replied:
Not add_car is ready to invalid tariff and full lot but can it fails for any other reason?

arturoaviles replied:
Fixed. I changed the error code to 507 (Insufficient Storage) when the Parking Lot is full

It could fail for many other reasons, some that I can think of are:

- Missing params while adding a car (I made the params required)
-  Endpoint params too large (if the user wants to add a car with a super large string as its license_plate for example). This can be solved by limiting the # of characters allowed per param

Other stuff that could be improved:

- Car license plate sanitization (Example: If the user sends an empty string (""), a string with spaces ("  ") or not valid license plate characters (these would need to be defined) and could be solved with a REGEX.

Another thing that I can think about is Duplicated license plates (But I think there are some countries that let the user choose the license plate string (What would happen if two users have their cars with the same license plate string? Example: "The #1". If duplication is not allowed, I would change the Keys used in my Parking Lot dictionary from positions (int) to the license plates (str). If the license plate is already in the Parking Lot, return an error. The /remove endpoint would need to be changed to retrieve the car by its license plate instead of the position.

kingdomik commented on app/main.py:
```diff
@@ -0,0 +1,109 @@
+				"status": "error",
+				"description": str(e)
+			},
+			status_code=200
```
Why status 200 if removing failed?

arturoaviles replied:
Fixed. Changed to status 404 Car not found.

kingdomik commented on app/models/models.py:
```diff
@@ -0,0 +1,151 @@
+		Returns:
+			timedelta: Total time for the ticket.
+		"""
+		try:
+			start = datetime.strptime(self.start, DATE_TIME_FORMAT)
+			finish = datetime.strptime(self.finish, DATE_TIME_FORMAT)
+		except ValueError as e:
+			raise ValueError("Invalid date format") from e
```
Do we really need to convert timestamp to string to later parse it?

arturoaviles replied:
Fixed. Datetimes are only converted to a string when sending back the JSON response because a datetime object is not JSON serializable.

kingdomik commented on app/models/models.py:
```diff
@@ -0,0 +1,151 @@
+			ParkingLotFullError: If the Parking Lot is full.
+		"""
+
+		try:
+			available_spot = self.available_spots.pop()
+		except KeyError as e: # Full parking Lot
+			raise ParkingLotFullError("No free space") from e
```
It is non-trivial that KeyError means no free space. Is it possible to improve this?

arturoaviles replied:
Yes. I improved it by doing it more "pythonic":

```python
available_spot = len(self.occupied_spots) + 1
if available_spot <= self.total_spots:
	base_tariff_cost = self.time_frame_tariffs.get(tariff)
	self.occupied_spots[available_spot] = Ticket(car.license_plate, tariff, available_spot, base_tariff_cost)
	return self.occupied_spots[available_spot]
raise ParkingLotFullError("No free space")
```

For this case, I made an improvement in my Parking Lot logic. I removed the set that I was using to get the available spots. Instead, I use the dictionary length + 1 to get the next spot to use if it is <= to TOTAL SPOTS.

kingdomik commented on app/models/models.py:
```diff
@@ -0,0 +1,151 @@
+
+		self.occupied_spots[available_spot] = Ticket(car.license_plate, tariff, available_spot)
+
+		return self.occupied_spots[available_spot]
```
What if we get an exception after `available_spots.pop()` and before `occupied_spots[available_spot] = ...`?

arturoaviles replied:
I improved the Parking Lot logic to address this sync issue. Now I'm only using 1 data structure to manage the cars in the parking lot. 

```
self.occupied_spots = {}
self.available_spots = {} (REMOVED)
```

Commit:
Correct error status codes:

- 404: When car not found in parking lot location
- 507: When parking lot storage is "full" (Insufficient Storage)

Commit:
Fix: Bug when adding car

An exception could occur while getting an available spot and parking the car

Commit:
Add pagination in parking lot list cars

Commit:
Add tariff cost changer endpoint

The owner of the parking lot can change the tariffs by sending the new costs to the /config endpoint

Commit:
Improve ticket datetimes management
